dawn/docs/parallel_model.rst

```diff
@@ -0,0 +1,251 @@
+   output fields we duplicate the writes which can have 2 side-effects
+   if neighboring blocks are out of sync: Block 1 writes in its extended
+   domain (line 2), block 2 writes (line 2) (same value), then reads
+   offset (line 3), then writes in the center (line 4). Now Block 1
+   reads (line 3) and gets the wrong value. This means that if we want
+   to solve the problem we need a global sync in that case (meaning
+   breaking the kernel call).
+
+.. code:: c++
+
+   1 vertical_region(k_start, k_end){
+   2   one = 1
+   3   two = one[i - 1]
+   4   one = 2
+   5 }
+
+translates to
+
+::
+
+   MultiStage_0 [parallel]
+       {
+         Stage_0
+         {
+           Do_0 { Start : End }
+           {
+             one[<no_horizontal_offset>,0] = int_type 1;
+           }
+           Extents: [(-1,0),(0,0),(0,0)]
+         }
+         Stage_1
+         {
+           Do_0 { Start : End }
+           {
+             two[<no_horizontal_offset>,0] = one[-1,0,0];
+           }
+           Extents: [<no_horizontal_extent>,(0,0)]
+         }
+         Stage_2
+         {
+           Do_0 { Start : End }
+           {
+             one[<no_horizontal_offset>,0] = int_type 2;
+           }
+           Extents: [<no_horizontal_extent>,(0,0)]
+         }
+       }
+
+--------------
+
+Compute Domain
+~~~~~~~~~~~~~~
+
+If the vertical region does not specify any horizontal extent, dawn
+ensures that the compute-domain of statements that others depend on is
+extended to fill all the intermediate fields as required. If a
+horizontal compute-domain is specified, it is never extended.
+
+Internal Contract of Data Structures
+------------------------------------
+
+StencilInstantiation
+~~~~~~~~~~~~~~~~~~~~
+
+StencilInstantiations (SI) are user defined and are never changed during
+any optimization. Each SI will be globally split from every other one
+and there is no optimization across SIs. For every SI, a callable API
+will be generated.
+
+Stencil
+~~~~~~~
+
+Stencils serve no purpose right now
+
+--------------
+
+   It might be useful to remove them altogether since even if they
+   existed, no concept can't be captured by the other structures
+
+--------------
+
+Multistage
+~~~~~~~~~~
+
+A Multistage (MS) is the equivalent of a global synchronization between
+all processes on the full domain. each Multistage can only have one loop
+order
+
+-  in the CUDA back end they are the equivalent of a kernel call
+-  in the GridTools back end they are the equivalent of a multistage
+-  in the naive back end every storage will be synchronized on the host
+   at the beginning of each multistage
+
+--------------
+
+   Can we discuss why we still have the loop order on this level? Would
+   it potentially be interesting to promote this to the Do level and
+   change the fusion strategy if we use the GridTools back end? Or do we
+   know that we are losing performance in the CUDA back end if we mix
+   forward and backward loops versus new kernel calls?
```

Comment:
I think I agree that we could put it into the Do-level. Then the only purpose of the multistage would be to resolve the write-after-read-after-write problem?

Reply:
we could let the do-method hold all the information about the k-loop, right? Then the concept of the MS would just be global sync.

Does that make sense?

Reply:
Yes. This doesn't change the fact that the multistage is at the point where the k-loop(s) happen, but the actual indices this is going to iterate over is inside the do-method. But I like the idea to have one responsibility per concept (global sync).

Reply:
Just an addition here from the discussion with @twicki : The global sync is needed for the write-after-read-after-write problem, but it is also needed for vertical-diagonal offsets (i+1, k+1), if `ij` and `k` loops may be exchanged. So if we have vertical-diagonal offsets, we need to split into two multistages.